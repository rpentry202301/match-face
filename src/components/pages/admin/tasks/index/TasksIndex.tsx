import OrangeButton from "@/components/ui/button/OrangeButton";
import SearchByJobs from "./parts/SearchByJobs";
import TaskList from "./parts/TaskList";
import Link from "next/link";
import { headers } from "next/headers";
import { TasksType } from "@/types/admin/tasks/types";

/**
 * @author Hayato Kobayashi
 * @todo_1 レスポンシブ対応
 */
const TasksIndex = async () => {
  // クエリから検索条件{ searchKeyword, departmentId }を取得
  const headerList = headers();
  const searchKeyword = headerList.get("searchKeyword");
  const departmentId = headerList.get("departmentId");

  // tasksデータの取得
  const res = await fetch(
    `${process.env.NEXT_API_URL}/api/admin/tasks?searchKeyword=${searchKeyword}&departmentId=${departmentId}`,
    {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
  const tasks: TasksType[] = await res.json();
  console.log("tasks", tasks[0].answerUserList.length);

  return (
    <main>
      <div className="mb-8">
        <SearchByJobs />
      </div>
      <div className="flex justify-center mb-8">
        <Link href={"/admin/tasks/register"} data-testid="link_task_register">
          <OrangeButton label="新規タスク作成" />
        </Link>
      </div>
      <div className="flex justify-center">
        <TaskList tasks={tasks} />
      </div>
    </main>
  );
};

export default TasksIndex;
