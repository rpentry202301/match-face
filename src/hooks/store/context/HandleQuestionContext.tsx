"use client";

import React, { useContext } from "react";

type Refine = {
  word: string[];
  department: number[];
};

const SelectContext = React.createContext<[Refine, React.Dispatch<React.SetStateAction<Refine>>] | undefined>(undefined);

export function SelectProvider({ children }: { children: React.ReactNode }) {
  const [select, setSelect] = React.useState<Refine>({
    word: [],
    department: [],
  });

  return (
    <SelectContext.Provider value={[select, setSelect]}>
      {children}
    </SelectContext.Provider>
  );
}

export function useRefine() {
  const context = useContext(SelectContext);
  if (context === undefined) {
    throw new Error("useRefine must be used within a SelectProvider");
  }
  return context;
}
