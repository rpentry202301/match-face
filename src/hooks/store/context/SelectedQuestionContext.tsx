"use client";
import {
  Dispatch,
  ReactNode,
  Reducer,
  createContext,
  useContext,
  useReducer,
} from "react";
import { QusetionsReducer } from "../reducer/selecterReducer";
import { QuestionsReducerAction } from "@/types/admin/tasks/register/types";

export type Questions = {
  projectId: number;
  list: {
    id: number;
    name: string;
  }[];
};

const SelectedQuestionContext = createContext<
  [Questions, Dispatch<QuestionsReducerAction>] | undefined
>(undefined);

export const SelectedQuestionProvider = ({
  children,
}: {
  children: ReactNode;
}) => {
  const [state, dispatch] = useReducer<
    Reducer<Questions, QuestionsReducerAction>
  >(QusetionsReducer, { projectId: 0, list: [] });
  return (
    <SelectedQuestionContext.Provider value={[state, dispatch]}>
      {children}
    </SelectedQuestionContext.Provider>
  );
};

export const useSelectedQuestion = () => {
  const context = useContext(SelectedQuestionContext);
  if (context === undefined)
    throw new Error("Function 'useSelectedQuestion' must be used in Provider.");
  return context;
};
