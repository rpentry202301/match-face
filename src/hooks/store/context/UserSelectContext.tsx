"use client";
import {
  Dispatch,
  ReactNode,
  Reducer,
  createContext,
  useContext,
  useReducer,
} from "react";
import { UsersReducer } from "../reducer/selecterReducer";
import { UsersReducerAction } from "@/types/admin/tasks/register/types";

export type Users = {
  id: string;
  name: string;
};

const UserSelectContext = createContext<
  [Users[], Dispatch<UsersReducerAction>] | undefined
>(undefined);

export const UserSelectProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer<Reducer<Users[], UsersReducerAction>>(
    // selecterReducer,
    UsersReducer,
    []
  );
  return (
    <UserSelectContext.Provider value={[state, dispatch]}>
      {children}
    </UserSelectContext.Provider>
  );
};

export const useUserSelect = () => {
  const context = useContext(UserSelectContext);
  if (context === undefined)
    throw new Error("Function 'useUserSelect' must be used in Provider.");
  return context;
};
