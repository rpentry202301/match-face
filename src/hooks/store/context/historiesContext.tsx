'use client'
import { createContext, Dispatch, useContext, useReducer } from "react";
import { SelectHistoryAction } from "@/types/admin/histories/admin_histories";
import { SelectHistoryReducer } from "../reducer/historiesReducer";

const SelectHistoryContext = createContext<{month:string,department:number[],skills:number[]}| undefined>(undefined);
const SelectHistoryDispatchContext = createContext<Dispatch<SelectHistoryAction>|undefined>(undefined)
const initialSelect = {month:"",department:[],skills:[]}

export function SelectHistoryProvider ({children}:{children:React.ReactNode}){
    const [state,dispatch] = useReducer(SelectHistoryReducer,initialSelect)
    return (
        <SelectHistoryContext.Provider value={state}>
            <SelectHistoryDispatchContext.Provider value={dispatch} >
            {children}
            </SelectHistoryDispatchContext.Provider>
        </SelectHistoryContext.Provider>
    )
}

export const useSelectHistory = () => {
    const context = useContext(SelectHistoryContext)
    if(context === undefined) {
        throw new Error("Function 'useSelectHistory' must be used in Provider.")
    }
    return context
}

export const useSelectHistoryDispatch = () => {
    const dispatch = useContext(SelectHistoryDispatchContext)
    if(dispatch === undefined) {
        throw new Error("Function 'useSelectHistoryDispatch' must be used in Provider.")
    }
    return dispatch
}
