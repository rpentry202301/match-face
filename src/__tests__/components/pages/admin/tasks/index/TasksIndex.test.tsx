import TasksIndex from "@/components/pages/admin/tasks/index/TasksIndex";
import {
  cleanup,
  fireEvent,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import { FilterProvider } from "@/hooks/store/context/TasksContext";

describe("TaskIndex.tsx", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    cleanup();
  });
  afterAll(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it("レンダリング時", () => {
    const { container } = render(
      <FilterProvider>
        <TasksIndex />
      </FilterProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it("新規タスク作成ボタンリンク", () => {
    render(
      <FilterProvider>
        <TasksIndex />
      </FilterProvider>
    );
    const LinkButton = screen.getByTestId("link_task_register");
    // console.log("LinkButton", LinkButton);
    expect(LinkButton).toHaveAttribute("href", "/admin/tasks/register");
  });
  describe("useFilter(context hook)のテスト", () => {
    beforeEach(() => {
      cleanup();
      render(
        <FilterProvider>
          <TasksIndex />
        </FilterProvider>
      );
    });
    it("フリーワード検索テスト", () => {
      // 案件名が"hoge", 職種が"Java"のデータを検索(tasks.id = 1)
      const input = screen.getByTestId("search-box") as HTMLInputElement;
      fireEvent.change(input, { target: { value: "hoge Java" } });
      fireEvent.click(screen.getByText("検索"));
      // id = 1 のtasksデータが検索結果に存在する
      expect(screen.queryByTestId("task-data-1")).toBeInTheDocument();
      // その他のデータが検索結果に存在しないかテスト
      expect(screen.queryByTestId("task-data-2")).toBeNull();
      expect(screen.queryByTestId("task-data-3")).toBeNull();
      expect(screen.queryByTestId("task-data-4")).toBeNull();
      expect(screen.queryByTestId("task-data-5")).toBeNull();
      expect(screen.queryByTestId("task-data-6")).toBeNull();
    });
    it("職種フィルターテスト", () => {
      // "Java", "FR"ボタンを押下→絞り込みボタン押下(tasks.id = 1, 3)
      fireEvent.click(screen.getByTestId("Java-btn"));
      fireEvent.click(screen.getByTestId("FR-btn"));
      fireEvent.click(screen.getByText("絞り込み"));
      // 職種が"Java", "FR"のデータが検索結果に存在する
      expect(screen.queryByTestId("task-data-1")).toBeInTheDocument();
      expect(screen.queryByTestId("task-data-3")).toBeInTheDocument();
      // その他の職種のデータが検索結果に存在しない
      expect(screen.queryByTestId("task-data-2")).toBeNull();
      expect(screen.queryByTestId("task-data-4")).toBeNull();
      expect(screen.queryByTestId("task-data-5")).toBeNull();
      expect(screen.queryByTestId("task-data-6")).toBeNull();
    });
    it("フリーワード検索&職種フィルターテスト", () => {
      // 検索ボックスに案件名"hoge"を入力, 職種フィルターボタン"Java"押下(tasks.id = 1)
      // 検索
      const input = screen.getByTestId("search-box") as HTMLInputElement;
      fireEvent.change(input, { target: { value: "hoge Java" } });
      fireEvent.click(screen.getByText("検索"));
      // 職種フィルターボタン
      expect(screen.queryByTestId("task-data-1")).toBeInTheDocument();
      // id = 1 のtasksデータが検索結果に存在する
      expect(screen.queryByTestId("task-data-1")).toBeInTheDocument();
      // その他のデータが検索結果に存在しないかテスト
      expect(screen.queryByTestId("task-data-2")).toBeNull();
      expect(screen.queryByTestId("task-data-3")).toBeNull();
      expect(screen.queryByTestId("task-data-4")).toBeNull();
      expect(screen.queryByTestId("task-data-5")).toBeNull();
      expect(screen.queryByTestId("task-data-6")).toBeNull();
    });
  });
});
