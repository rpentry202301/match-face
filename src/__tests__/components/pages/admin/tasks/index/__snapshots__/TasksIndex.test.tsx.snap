// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskIndex.tsx レンダリング時 1`] = `
<div>
  <main>
    <div
      class="mb-8"
    >
      <div
        class="flex flex-col items-center border-2 rounded-md w-3/6 mx-auto mt-6 p-3"
      >
        <div>
          <form
            class="flex items-center mb-4"
          >
            <div>
              <input
                class="border border-light-gray text-xs p-1 w-96"
                data-testid="search-box"
                id="search"
              />
            </div>
            <button
              class="bg-white hover:bg-gray-100 active:bg-gray-200 active:translate-y-0.5 active:shadow-sm py-2 px-4 rounded-full border shadow-md text-xs ml-2"
              type="submit"
            >
              検索
            </button>
          </form>
        </div>
        <div
          class="flex items-center mb-4"
        >
          <div
            class="mx-3"
          >
            <button
              class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
              data-testid="Java-btn"
            >
              Java
            </button>
          </div>
          <div
            class="mx-3"
          >
            <button
              class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
              data-testid="PHP-btn"
            >
              PHP
            </button>
          </div>
          <div
            class="mx-3"
          >
            <button
              class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
              data-testid="FR-btn"
            >
              FR
            </button>
          </div>
          <div
            class="mx-3"
          >
            <button
              class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
              data-testid="CL-btn"
            >
              CL
            </button>
          </div>
          <div
            class="mx-3"
          >
            <button
              class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
              data-testid="ML-btn"
            >
              ML
            </button>
          </div>
          <div
            class="mx-3"
          >
            <button
              class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
              data-testid="QA-btn"
            >
              QA
            </button>
          </div>
        </div>
        <div>
          <div>
            <button
              class="bg-orange drop-shadow-md hover:saturate-150 active:drop-shadow-none active:shadow-inner hover:bg-depp-orange active:bg-depp-orange active:translate-y-0.5 text-white rounded-xl p-2 h-auto w-28 text-sm"
            >
              絞り込み
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex justify-center mb-8"
    >
      <a
        data-testid="link_task_register"
        href="/admin/tasks/register"
      >
        <div>
          <button
            class="bg-orange drop-shadow-md hover:saturate-150 active:drop-shadow-none active:shadow-inner hover:bg-depp-orange active:bg-depp-orange active:translate-y-0.5 text-white rounded-xl p-2 w-40 h-auto text-lg"
          >
            新規タスク作成
          </button>
        </div>
      </a>
    </div>
    <div
      class="flex justify-center"
    >
      <div>
        <table
          class="border-collapse border-2 border-deep-gray text-sm"
        >
          <thead
            class="bg-light-gray"
          >
            <tr>
              <th
                class="border-2 border-deep-gray p-2 text-center w-72"
              >
                案件名
              </th>
              <th
                class="border-2 border-deep-gray p-2 text-center w-16"
              >
                質問数
              </th>
              <th
                class="border-2 border-deep-gray p-2 text-center w-16"
              >
                職種
              </th>
              <th
                class="border-2 border-deep-gray p-2 text-center w-72"
              >
                回答ユーザー
              </th>
              <th
                class="border-2 border-deep-gray p-2 text-center w-28"
              >
                期日
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              data-testid="task-data-1"
            >
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                hoge
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                3
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                Java
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                佐藤太郎、鈴木次郎、伊藤三郎
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                2100-01-01
              </td>
            </tr>
            <tr
              data-testid="task-data-2"
            >
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                foo
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                3
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                PHP
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                佐藤太郎、鈴木次郎、伊藤三郎...
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                2100-01-01
              </td>
            </tr>
            <tr
              data-testid="task-data-3"
            >
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                bar
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                3
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                FR
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                佐藤太郎、鈴木次郎、伊藤三郎
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                2100-01-01
              </td>
            </tr>
            <tr
              data-testid="task-data-4"
            >
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                hogehoge
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                3
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                CL
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                佐藤太郎、鈴木次郎、伊藤三郎
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                2200-01-01
              </td>
            </tr>
            <tr
              data-testid="task-data-5"
            >
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                foofoo
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                3
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                ML
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                佐藤太郎、鈴木次郎、伊藤三郎...
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                2200-01-01
              </td>
            </tr>
            <tr
              data-testid="task-data-6"
            >
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                barbar
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                3
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                QA
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                佐藤太郎、鈴木次郎、伊藤三郎
              </td>
              <td
                class="border-2 border-deep-gray p-2 text-center"
              >
                2200-01-01
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </main>
</div>
`;
